File: src/main/java/fi/finlit/edith/ui/mixins/TermAutocomplete.java
package fi.finlit.edith.ui.mixins;

import java.util.List;

import org.apache.commons.lang.StringUtils;
import org.apache.tapestry5.annotations.Import;
import org.apache.tapestry5.json.JSONArray;
import org.apache.tapestry5.json.JSONObject;
import org.apache.tapestry5.services.javascript.JavaScriptSupport;

import fi.finlit.edith.domain.Term;

// TODO : use common superclass for TermAutocomplete and JQueryAutocomplete
@Import(library = { "term-autocomplete.js" })
public class TermAutocomplete extends JQueryAutocomplete {

    @Override
    protected void init(String elementId, String ajaxURI, JavaScriptSupport renderSupport) {
        super.init(elementId, ajaxURI, renderSupport);
        // Hooking our own function
        renderSupport.addInitializerCall("termAutocompleter",
                new JSONObject("elementId", elementId));
    }

    @Override
    protected JSONArray generateResponse(List<Object> matches) {
        JSONArray a = new JSONArray();

        for (Object o : matches) {
            Term term = (Term) o;
            /* "value" is the item used by the autocompletion to visualize the list element. */
            JSONObject json = new JSONObject().put("basicForm", term.getBasicForm());
            json.put("language", String.valueOf(term.getLanguage()));
            json.put("meaning", term.getMeaning());
            json.put("id", term.getId());
            if (term.getMeaning() != null) {
                json.put("value",
                        term.getBasicForm() + " - " + StringUtils.abbreviate(term.getMeaning(), 32));
            } else {
                json.put("value", term.getBasicForm());
            }
            a.put(json);
        }

        return a;
    }

}

File: src/main/scala/fi/finlit/edith/ui/mixins/TermAutocomplete.scala
package fi.finlit.edith.ui.mixins

import java.util.List
import org.apache.commons.lang.StringUtils
import org.apache.tapestry5.annotations.Import
import org.apache.tapestry5.json.JSONArray
import org.apache.tapestry5.json.JSONObject
import org.apache.tapestry5.services.javascript.JavaScriptSupport
import fi.finlit.edith.sql.domain.Term
//remove if not needed
import scala.collection.JavaConversions._

@Import(library = ("term-autocomplete.js"))
class TermAutocomplete extends JQueryAutocomplete {

  protected override def init(elementId: String, ajaxURI: String, renderSupport: JavaScriptSupport) {
    super.init(elementId, ajaxURI, renderSupport)
    renderSupport.addInitializerCall("termAutocompleter", elementId)
  }

  protected override def generateResponse(matches: List[Any]): JSONArray = {
    var a = new JSONArray()
    for (o <- matches) {
      var term = o.asInstanceOf[Term]
      var json = new JSONObject().put("basicForm", term.getBasicForm)
      json.put("language", String.valueOf(term.getLanguage))
      json.put("meaning", term.getMeaning)
      json.put("id", term.getId)
      if (term.getMeaning != null) {
        json.put("value", term.getBasicForm + " - " + StringUtils.abbreviate(term.getMeaning, 32))
      } else {
        json.put("value", term.getBasicForm)
      }
      a.put(json)
    }
    a
  }
}
